### ISSUE TITLE
[Perf] Regression in Linq Order Benchmarks

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#49160]
### Run Information

Architecture | arm64
-- | --
OS | ubuntu 18.04
Baseline | [d5ab93c4a8e45da2dab8924c2165000bf78f84e6](https://github.com/dotnet/runtime/commit/d5ab93c4a8e45da2dab8924c2165000bf78f84e6)
Compare | [cbb5b905d67a91ecf293c8807414b7791713b02d](https://github.com/dotnet/runtime/commit/cbb5b905d67a91ecf293c8807414b7791713b02d)
Diff | [Diff](https://github.com/dotnet/runtime/compare/8b391f60276136954a17fb3dc1f254fb98d82022...0cbfb7459ba8e8f094b57ab0403726112c786a79)

### Regressions in LinqBenchmarks

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Order00LinqQueryX](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/master_arm64_ubuntu 18.04/LinqBenchmarks.Order00LinqQueryX.html>) | 226.12 ms | 270.32 ms | 1.20 | | |
[Order00LinqMethodX](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/master_arm64_ubuntu 18.04/LinqBenchmarks.Order00LinqMethodX.html>) | 230.02 ms | 269.94 ms | 1.17 | | |
[Order00ManualX](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/master_arm64_ubuntu 18.04/LinqBenchmarks.Order00ManualX.html>) | 302.15 ms | 411.12 ms | 1.36 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_02_2021/refs/heads/master_arm64_ubuntu%2018.04_Regression/LinqBenchmarks_1.png>)
![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_02_2021/refs/heads/master_arm64_ubuntu%2018.04_Regression/LinqBenchmarks_2.png>)
![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_02_2021/refs/heads/master_arm64_ubuntu%2018.04_Regression/LinqBenchmarks_3.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_02_2021/refs/heads/master_arm64_ubuntu%2018.04_Regression/LinqBenchmarks.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'LinqBenchmarks*'
```
<details>

.
### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-91a5042a-d43a-4b77-871e-3b07890addc1c65b02e190f479b86/f37209d8-951d-4fdb-b742-f229ea2c60a7.zip?sv=2019-07-07&se=2021-03-25T12%3A34%3A32Z&sr=c&sp=rl&sig=8S7zkwzMsy7w8eFWTOR3Xs69A0eSSmN3HIBzXLfUuRE%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-1a29b912-dd21-4552-bf91-409f6debb6a9fc990de6ca8475a80/746eb388-7245-4416-ae40-1cada4fa6f1e.zip?sv=2019-07-07&se=2021-03-26T22%3A58%3A43Z&sr=c&sp=rl&sig=ruXse3asX%2BK4oMC4NpdJZH%2Fm%2BRHrKxgQUmtg%2Fsd7P40%3D>)

#### LinqBenchmarks.Order00LinqQueryX

```log
[219204516.970 ; 226805981.622) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[226805981.622 ; 235528812.551) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[235528812.551 ; 242930455.849) | @@@@@@@@@@@@@
[242930455.849 ; 251375191.122) | 
[251375191.122 ; 259773457.812) | @
[259773457.812 ; 267069172.458) | @@@@@@@@@@
[267069172.458 ; 275311914.563) | @@@@@@@@@@@@@@@
```
#### LinqBenchmarks.Order00LinqMethodX

```log
[219293941.975 ; 226860651.666) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[226860651.666 ; 235455653.528) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[235455653.528 ; 242550281.936) | @@@@@
[242550281.936 ; 249269681.504) | 
[249269681.504 ; 255654555.778) | 
[255654555.778 ; 265249408.198) | @@@@@@@@@@@@@@@@@@
[265249408.198 ; 275092573.821) | @@@@@@@
```
#### LinqBenchmarks.Order00ManualX

```log
[284288503.463 ; 297088106.846) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[297088106.846 ; 309191472.854) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[309191472.854 ; 320867881.858) | @@@@@@@@@@@
[320867881.858 ; 336120864.083) | @
[336120864.083 ; 347792070.996) | @@@@@
[347792070.996 ; 360367994.551) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[360367994.551 ; 372471360.559) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[372471360.559 ; 386473406.970) | @@@@@@@@@@@@@@@
[386473406.970 ; 398482828.351) | @@@@@
[398482828.351 ; 417169873.271) | @@@@@@@@@
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by eiriktsarpalis** [comment]
I initially suspected https://github.com/dotnet/runtime/commit/c63f06cd1635b935ead3c3fbc33a52221bc79bb6, but it seems that the change predates the baseline commit. I ran a git-bisect which indicated that this commit https://github.com/dotnet/runtime/commit/e912151eda83eb13b41896fe480cd6f5f1465af2 is the source of the regression. It is not clear to me how it may affect [benchmark code](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/runtime/Linq/Linq.cs). cc @stephentoub 

```
git bisect start
# bad: [cbb5b905d67a91ecf293c8807414b7791713b02d] Make Exception.StackTrace resilient to missing attribute assemblies (#48535)
git bisect bad cbb5b905d67a91ecf293c8807414b7791713b02d
# good: [d5ab93c4a8e45da2dab8924c2165000bf78f84e6] Fix ArgIterator::GetNextOffset for macOS arm64 (#48632)
git bisect good d5ab93c4a8e45da2dab8924c2165000bf78f84e6
# good: [ecfcdfd776da13f12f521e3b5b9a13483a0a5038] Support SPMI collection of tiered compilation runs, fix utf8 issues (#48665)
git bisect good ecfcdfd776da13f12f521e3b5b9a13483a0a5038
# good: [8f0f819d49c75a57dd2bc3e0a80e7cbd8ec520e1] Updating snaps (#48668)
git bisect good 8f0f819d49c75a57dd2bc3e0a80e7cbd8ec520e1
# good: [75065d1d521bb57381a20e6898e58df7f48ccc57] R2RTest cleanup for Serp compile (#48528)
git bisect good 75065d1d521bb57381a20e6898e58df7f48ccc57
# good: [eb03e0f7bc396736c7ac59cf8f135d7c632860dd] Fixup richnav triggers (#48714)
git bisect good eb03e0f7bc396736c7ac59cf8f135d7c632860dd
# bad: [e912151eda83eb13b41896fe480cd6f5f1465af2] Replace non-generic WeakReference usage in several libraries (#48618)
git bisect bad e912151eda83eb13b41896fe480cd6f5f1465af2
# good: [dd6bc8ce6bfc6832c33800200d402f8913b8cfa2] Enhance coreclr processing of event pipe configuration to be able to write the pid of the current process into the nettrace file (#48537)
git bisect good dd6bc8ce6bfc6832c33800200d402f8913b8cfa2
# first bad commit: [e912151eda83eb13b41896fe480cd6f5f1465af2] Replace non-generic WeakReference usage in several libraries (#48618)
# good: [dd6bc8ce6bfc6832c33800200d402f8913b8cfa2] Enhance coreclr processing of event pipe configuration to be able to write the pid of the current process into the nettrace file (#48537)
git bisect good dd6bc8ce6bfc6832c33800200d402f8913b8cfa2
# first bad commit: [e912151eda83eb13b41896fe480cd6f5f1465af2] Replace non-generic WeakReference usage in several libraries (#48618)
```

**Comment by stephentoub** [comment]
> I ran a git-bisect which indicated that this commit e912151 is the source of the regression

You're sure?  I can take a look, but it's hard to see how that could have any possible impact.  I don't think any of the touched library code is used by either the LINQ API in question nor by the test itself.

**Comment by eiriktsarpalis** [comment]
@stephentoub please disregard, I hadn't noticed this was an arm64 specific regression. Whatever performance regression I did measure is probably unrelated to this report.

**Comment by eiriktsarpalis** [comment]
The particular report seems to contain no information past February, here's some more recent data:

https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_ubuntu%2018.04/LinqBenchmarks.Order00LinqQueryX.html
https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_ubuntu%2018.04/LinqBenchmarks.Order00LinqMethodX.html
https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_ubuntu%2018.04/LinqBenchmarks.Order00ManualX.html

![image](https://user-images.githubusercontent.com/2813363/129441713-c134bf1c-1d4d-4e3f-9ef1-436640a81334.png)

Based on the above, it would seem like the performance regression was fixed by #48634.
